src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdBackgroundTask.cs

```diff
@@ -0,0 +1,114 @@
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using OrchardCore.BackgroundTasks;
+using OrchardCore.OpenId.Abstractions.Models;
+using OrchardCore.OpenId.Services.Managers;
+
+namespace OrchardCore.OpenId.Services
+{
+    [BackgroundTask(Group = "OrchardCore.OpenId")]
+    public class OpenIdBackgroundTask : IBackgroundTask
+    {
+        private readonly ILogger<OpenIdBackgroundTask> _logger;
+
+        public OpenIdBackgroundTask(ILogger<OpenIdBackgroundTask> logger)
+        {
+            _logger = logger;
+        }
+
+        public async Task DoWorkAsync(IServiceProvider serviceProvider, CancellationToken cancellationToken)
+        {
+            // Note: this background task is responsible of automatically removing orphaned tokens/authorizations
+            // (i.e tokens that are no longer valid and ad-hoc authorizations that have no valid tokens associated).
+            // Since ad-hoc authorizations and their associated tokens are removed as part of the same operation
+            // when they no longer have any token attached, it's more efficient to remove the authorizations first.
+
+            await DeleteAuthorizationsAsync();
+            await DeleteTokensAsync();
+
+            async Task DeleteAuthorizationsAsync()
+            {
+                // Note: the authorization manager MUST be resolved from the scoped provider
+                // as it depends on scoped stores that should be disposed as soon as possible.
+                var manager = serviceProvider.GetRequiredService<OpenIdAuthorizationManager>();
+
+                ImmutableArray<IOpenIdAuthorization> authorizations;
+
+                do
+                {
+                    // Note: don't use an offset here, as the elements returned by this method
+                    // are progressively removed from the database immediately after calling it.
+                    authorizations = await manager.ListInvalidAsync(100, 0, cancellationToken);
```

Comment:
Note: I have no strong opinion about the number of items we should retrieve by call. I'm fine changing it if necessary.